playbooks/migrating-postgres-with-rubyrep.md

```diff
@@ -0,0 +1,227 @@
+ALTER TABLE bar DROP CONSTRAINT fk_rails_80eb82ccbf;
+```
+
+5) Create the file `hokusai/rubyrep-config.yml` (Credentials here are redacted but they should be supplied to Kubernetes and this file should not be checked in!) The "left" configuration refers to the source database and the "right" refers to the target database.  Note that for rails apps we can ignore both `ar_internal_metadata` and `schema_migrations` tables.
+
+```
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    app: myapp
+  name: rubyrep-myapp
+  namespace: default
+data:
+  default.conf: |
+    RR::Initializer::run do |config|
+      config.left = {
+        :adapter  => 'postgresql',
+        :database => 'REDACTED',
+        :username => 'REDACTED',
+        :password => 'REDACTED',
+        :host     => 'OLD_DATABASE',
+        :sslmode  => 'require',
+        :logger   => STDOUT
+      }
+
+      config.right = {
+        :adapter  => 'postgresql',
+        :database => 'REDACTED',
+        :username => 'REDACTED',
+        :password => 'REDACTED',
+        :host     => 'NEW_DATABASE',
+        :sslmode  => 'require',
+        :logger   => STDOUT
+      }
+
+      config.include_tables /./
+      config.exclude_tables 'ar_internal_metadata'
+      config.exclude_tables 'schema_migrations'
+
+      config.options[:adjust_sequences] = false
+
+      config.options[:logged_replication_events] = [:all_changes, :all_conflicts]
+
+    end
+```
+
+Create the configmap with:
+```
+$ hokusai [staging|production] create --filename ./hokusai/rubyrep-config.yml
+```
+
+6) Create the file `./hokusai/rubyrep.yml`.  When the deployment is created, rubyrep will perform an initial sync.
+
+```
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: myapp-rubyrep
+  namespace: default
+spec:
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 0
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app: myapp
+        component: rubyrep
+        layer: data
+      name: myapp-rubyrep
+    spec:
+      containers:
+        - name: rubyrep-myapp
+          image: artsy/rubyrep
+          args: ["/rubyrep-2.0.1/rubyrep", "--verbose", "replicate", "-c", "/mnt/default.conf"]
+          imagePullPolicy: Always
+          resources:
+            requests:
+              cpu: 200m
+              memory: 256Mi
+          volumeMounts:
+            - name: rubyrep-myapp
+              mountPath: /mnt/default.conf
+              subPath: default.conf
+      volumes:
+        - name: rubyrep-myapp
+          configMap:
+            name: rubyrep-myapp
+```
+
+Create the deployment with:
+```
+$ hokusai [staging|production] create --filename ./hokusai/rubyrep.yml
+```
+
+7) When the database is synced and replication is active, we are ready to make a cutover. To avoid race conditions regarding auto-incrementing primary keys, we will have to accept a brief window of downtime while performing a cutover to the new database.
```

Comment:
Could we avoid the downtime and potential conflicts by increasing the id sequences of the new database (to sufficiently high values) in advance of the replication? Replication could continue to insert new data with the lower values, and any direct inserts after cut-over could pick up at the higher value.

Reply:
I believe Rubyrep inserts rows _without_ making use of "autoincrement" so this may work if we are okay with skipping a few primary key ids.

@joeyAghion so you are suggesting that if, say `messagers_id_seq` is at `2000` we bump it to `2100` before cutover, allowing for any replicated messages to fill up to 2100 and any new messages to start from 2100?

A note on sequences - we are explicitly disabling Rubyrep's "adjust sequences" [config parameter](https://github.com/rubyrep/rubyrep/blob/master/lib/rubyrep/replication_initializer.rb#L63) as it will try and reorder sequences in both databases (i.e. left database would end up with only even primary keys and right database would end up with only odd) - note in the code the line "The sequence is always updated in both databases".

See http://www.rubyrep.org/configuration.html for sequence configuration options but I would hesitate to have rubyrep make any destructive changes to the source database.

Reply:
Yup that was my thought, but I don't know what's safest.

Reply:
As long as it's easy to configure the right ID offset for all tables this should work... we would have to choose an arbitrary offset but we can just set it cautiously high for all tables based on what we expect the highest throughput table to experience.

Reply:
Thanks for the feedback @cavvia - since we don't have datadog monitoring set up I'll poll the production tables and report avg write throughput / minute here.

Reply:
Impulse production seems to have a max throughput on the messages table of ~6 writes / minute